renderer/FrameBuffer.cpp

```diff
@@ -464,3 +464,221 @@ void FB_TogglePrimary( bool on ) {
+		frameBuffers.allFrameBuffers.push_back( frameBuffers.backBuffer );
+		common->Printf( "Backbuffer dimensions: %dx%d\n", frameBuffers.backBuffer->width, frameBuffers.backBuffer->height );
+	}
+	return frameBuffers.backBuffer;
+}
+
+void FrameBuffer::CreateColorBuffer() {
+	if( colorBuffer == 0 ) {
+		qglGenRenderbuffers( 1, &colorBuffer );
+		if( colorBuffer == 0 ) {
+			common->FatalError( "Failed to allocate color render buffer" );
+		}
+	}
+	qglBindRenderbuffer( GL_RENDERBUFFER, colorBuffer );
+	if( msaa > 1 ) {
+		qglRenderbufferStorageMultisample( GL_RENDERBUFFER, msaa, GL_RGBA, width, height );
+	} else {
+		qglRenderbufferStorage( GL_RENDERBUFFER, GL_RGBA, width, height );
+	}
+
+	Bind();
+	qglFramebufferRenderbuffer( GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_RENDERBUFFER, colorBuffer );
+}
+
+void FrameBuffer::CreateDepthStencilBuffer() {
+	if( depthStencilBuffer == 0 ) {
+		qglGenRenderbuffers( 1, &depthStencilBuffer );
+		if( depthStencilBuffer == 0 ) {
+			common->FatalError( "Failed to allocate depth/stencil render buffer" );
+		}
+	}
+	qglBindRenderbuffer( GL_RENDERBUFFER, depthStencilBuffer );
+	if( msaa > 1 ) {
+		qglRenderbufferStorageMultisample( GL_RENDERBUFFER, msaa, GL_DEPTH24_STENCIL8, width, height );
+	} else {
+		qglRenderbufferStorage( GL_RENDERBUFFER, GL_DEPTH24_STENCIL8, width, height );
+	}
+
+	Bind();
+	qglFramebufferRenderbuffer( GL_FRAMEBUFFER, GL_DEPTH_STENCIL_ATTACHMENT, GL_RENDERBUFFER, depthStencilBuffer );
+}
+
+void FrameBuffer::AddColorTexture( idImage *colorTexture ) {
+	if( msaa > 1 ) {
+		common->FatalError( "Rendering to texture for a framebuffer with AA enabled is not currently supported." );
+	}
+	if( colorBuffer != 0 ) {
+		common->Warning( "Setting framebuffer color attachment to texture, but a color render buffer was previously created." );
+	}
+	Bind();
+	qglFramebufferTexture2D( GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, colorTexture->texnum, 0 );
+}
+
+void FrameBuffer::AddDepthStencilTexture( idImage *depthStencilTexture ) {
+	if( msaa > 1 ) {
+		common->FatalError( "Rendering to texture for a framebuffer with AA enabled is not currently supported." );
+	}
+	if( depthStencilBuffer != 0 ) {
+		common->Warning( "Setting framebuffer depth/stencil attachment to texture, but a render buffer was previously created." );
+	}
+	Bind();
+	qglFramebufferTexture2D( GL_FRAMEBUFFER, GL_DEPTH_STENCIL_ATTACHMENT, GL_TEXTURE_2D, depthStencilTexture->texnum, 0 );
+}
+
+void FrameBuffer::AddDepthStencilTextures( idImage *depthTexture, idImage *stencilTexture ) {
+	if( msaa > 1 ) {
+		common->FatalError( "Rendering to texture for a framebuffer with AA enabled is not currently supported." );
+	}
+	if( depthStencilBuffer != 0 ) {
+		common->Warning( "Setting framebuffer depth/stencil attachment to texture, but a render buffer was previously created." );
+	}
+	Bind();
+	qglFramebufferTexture2D( GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_TEXTURE_2D, depthTexture->texnum, 0 );
```

Comment:
I think attached textures should be saved in Framebuffer somehow, not only in GL object. It makes it easier to debug, and add some possibilities.
For instance, you forbid attaching a color texture to FBO with color renderbuffer already created. But you do not forbid creating color renderbuffer when color texture is already attached. Is it intentional? It is hard to think that engine is operating properly when it first attaches a texture to FBO and then switches to renderbuffer for the rest of time...

Reply:
Fair point. I went for the simplest initial implementation, and since we don't make use of the textures after binding them, I omitted them.